### PR TITLE
add ens stuffz

### PR DESCRIPTION
**Pull request by jaycenhorton**
These are the prerequisites for consuming ENS inside of truffle develop/test. It deploys an ENS registry, creates a registrar, and assigns `nori.eth` so it can be consumed in migrations. This pr doesn't really consume it, and subsequent prs begin that journey, but probably arent perfect either

see: https://twitter.com/jaycenhorton/status/1015790432926568448

### CODE REVIEW

**Comment by jaycenhorton** [comment]
[Add ENS to truffle](https://trello.com/c/Nv8PK0Mk/379-add-ens-to-truffle)

File: contracts/ens/FIPSRegistrar.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
+pragma solidity ^0.4.24;
+
+import "./ENS.sol";
+
+/**
+ * @title FIPSRegistrar is a registrar that allocates subdomains to the first person to claim them.
+ */
+contract FIFSRegistrar {
```
Comment (pcardune) [inline contracts/ens/FIPSRegistrar.sol]: FIPS or FIFS?

File: migrations/1_configure_ens.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+const namehash = require('eth-ens-namehash');
+
+const ENS = artifacts.require('./ENSRegistry.sol');
+const FIFSRegistrar = artifacts.require('./FIFSRegistrar.sol');
```
Comment (pcardune) [inline migrations/1_configure_ens.js]: don't use file paths. just use class names.

File: test/helpers/ens.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+const getENSDetails = async (network, artifacts, web3) => {
+  let ens, resolver;
+  console.log(`Looking up existing registry at nori.test ENS on ${network}`);
+  if (network === 'ropstenGeth' || network === 'ropsten') {
+    ens = new ENS(web3.currentProvider);
+    resolver = ens.resolver('nori.test').addr();
+  } else {
+    try {
+      ens = await artifacts.require('./ENSRegistry.sol').deployed();
```
Comment (pcardune) [inline test/helpers/ens.js]: ditto

File: test/helpers/ens.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      resolver = ens.resolver(namehash.hash('nori.eth'));
+    } catch (e) {
+      console.log('make sure ENS is deployed first');
+    }
+  }
+  return artifacts.require('./RootRegistryV0_1_0.sol').at(await resolver);
```
Comment (pcardune) [inline test/helpers/ens.js]: ditto

File: contracts/ens/ENSRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,99 @@
+pragma solidity ^0.4.24;
+
+import "./ENS.sol";
+
+/**
+  * @title The ENS registry contract
+*/
+contract ENSRegistry is ENS {
```
Comment (pcardune) [inline contracts/ens/ENSRegistry.sol]: Are the ENS contracts not available in an existing repository somewhere? Did you hand code this or copy it from somewhere? Seems weird that we would have to have our own implementation of ENS in here.

File: contracts/ens/FIPSRegistrar.sol (hunk 1 above)
Comment (jaycenhorton) [reply contracts/ens/FIPSRegistrar.sol]: Fifs

Comment (jaycenhorton) [reply contracts/ens/FIPSRegistrar.sol]: will fix title note

File: contracts/ens/ENSRegistry.sol (hunk 5 above)
Comment (jaycenhorton) [reply contracts/ens/ENSRegistry.sol]: No hand code, stripped bare minimum of what I needed from https://github.com/ensdomains/ens/tree/master/contracts

**Comment by jaycenhorton** [comment]
The reason for this is because ALL existing ens implementations for local development are simply broken for truffle. They try spinning up multiple instances of testrpcs and propogate stack underflows inside truffles console. Ive notified all appropriate node module maintainers. For now this is the easiest solution.

Asking for you to trust me on that bit, as I spent a few hours of research on twittering with maintainers to verify
 https://twitter.com/jaycenhorton/status/1015790432926568448

**Comment by pcardune** [comment]
ok, that makes sense. Assuming those node modules are hosted on github, we should open github issues and link to them from here in case anyone else runs into the same issues.

**Pull request merged**